Mathlib/Algebra/Module/Submodule/Pointwise.lean

```diff
@@ -277,4 +289,236 @@ scoped[Pointwise] attribute [instance] Submodule.pointwiseMulActionWithZero
+      fun h => ⟨m, h, (one_smul _ _).symm⟩⟩
+  mul_smul s t x := le_antisymm
+    (set_smul_submodule_le _ _ _ <| by
+      rintro _ n ⟨a, b, ha, hb, rfl⟩ hn
+      rw [mul_smul]
+      apply mem_set_smul_submodule_of_mem_mem (mem1 := ha)
+      exact mem_set_smul_submodule_of_mem_mem (mem1 := hb) (mem2 := hn))
+    (set_smul_submodule_le _ _ _ <| by
+      rintro r m hr hm
+      have : SMulCommClass R R x := ⟨fun r s m => Subtype.ext <| show r • s • m.1 = s • r • m.1
+        from smul_comm _ _ _⟩
+      rw [mem_set_smul_submodule] at hm
+      obtain ⟨c, hc1, rfl⟩ := hm
+      delta Finsupp.sum
+      simp only [AddSubmonoid.coe_finset_sum, coe_toAddSubmonoid, SetLike.val_smul]
+      rw [Finset.smul_sum]
+      refine Submodule.sum_mem _ ?_
+      intro r' hr'
+      rw [← mul_smul]
+      apply mem_set_smul_submodule_of_mem_mem (mem2 := (c _).2)
+      exact ⟨r, r', hr, hc1 hr', rfl⟩)
+
+scoped[Pointwise] attribute [instance] Submodule.pointwiseSetMulActionSubmodule
+
+lemma set_smul_submodule_eq_iSup [SMulCommClass R R M] :
+    s • N = ⨆ (r ∈ s), r • N :=
+  set_smul_submodule_eq_of_le _ _ _
+    (fun r m hr hm => (show r • N ≤ _ by
+      rw [iSup_subtype', ← sSup_range]
+      exact le_sSup ⟨⟨r, hr⟩, rfl⟩) ⟨_, hm, rfl⟩) <| by
+    rw [iSup_subtype', ← sSup_range, sSup_le_iff]
+    rintro _ ⟨⟨x, hx⟩, rfl⟩ _ ⟨y, hy, rfl⟩
+    exact mem_set_smul_submodule_of_mem_mem (mem1 := hx) (mem2 := hy)
+
+/-- subset of a ring has a distributive multiplicative action on its submodules-/
+protected def pointwiseSetDistribMulActionSubmodule [SMulCommClass R R M] :
+    DistribMulAction (Set R) (Submodule R M) where
+  smul_zero s := set_smul_bot_submodule s
+  smul_add s x y := le_antisymm
+    (set_smul_submodule_le _ _ _ <| by
+      rintro r m hr hm
+      rw [add_eq_sup, Submodule.mem_sup] at hm
+      obtain ⟨a, ha, b, hb, rfl⟩ := hm
+      rw [smul_add, add_eq_sup, Submodule.mem_sup]
+      exact ⟨r • a, mem_set_smul_submodule_of_mem_mem (mem1 := hr) (mem2 := ha),
+        r • b, mem_set_smul_submodule_of_mem_mem (mem1 := hr) (mem2 := hb), rfl⟩)
+    (sup_le_iff.mpr ⟨set_smul_submodule_mono_right _ le_sup_left,
+      set_smul_submodule_mono_right _ le_sup_right⟩)
+
+scoped[Pointwise] attribute [instance] Submodule.pointwiseSetDistribMulActionSubmodule
+
+lemma sup_set_smul_submodule_eq_sup (s t : Set R) :
+    (s ⊔ t) • N = s • N ⊔ t • N :=
+  set_smul_submodule_eq_of_le _ _ _
+    (by rintro r n (hr|hr) hn
+        · exact Submodule.mem_sup_left (mem_set_smul_submodule_of_mem_mem _ _ hr hn)
+        · exact Submodule.mem_sup_right (mem_set_smul_submodule_of_mem_mem _ _ hr hn))
+    (sup_le
+      (set_smul_submodule_mono_left _ le_sup_left)
+      (set_smul_submodule_mono_left _ le_sup_right))
+
+lemma span_smul_submodule {R' M' : Type*} [CommSemiring R'] [AddCommMonoid M'] [Module R' M']
+    (s : Set R') (N : Submodule R' M') :
+    (Ideal.span s : Set R') • N = s • N :=
```

Comment:
I think there should instead be a lemma showing that the action is equal to `Submodule.hasSMul'` for ideals, and this is then just `Submodule.ideal_span_singleton_smul`.

Reply:
`s` is not necessarily a singleton ... I agree that `I • N = (I : Set R') • N` is a useful lemma, but I'm not sure it helps with the proof ...

Reply:
Do we not have `Ideal.span s • N = s • N`? We should also have such a lemma in any case.

Reply:
Has this been addressed?

Reply:
Is this lemma necessary, it is just two rewrites away
```lean
lemma span_smul_submodule'
    (s : Set R) (N : Submodule R M) :
    (Ideal.span s) • N = s • N := by
  rw [smul_eq_set_smul, Submodule.span_smul_submodule]
```

Reply:
Yes, because `smul_eq_set_smul` is an unfolding lemma.

Reply:
Where is `smul_eq_set_smul`? And I'd argue that `span_smul_submodule'` is necessary and `span_smul_submodule` isn't if you only want to choose one.

Reply:
And this should be called `span_set_smul_submodule` or `coe_span_set_submodule` if you want to keep it.

Reply:
Done, thanks for the naming suggestion.